src/Sylius/Component/Core/Resolver/DefaultPaymentMethodResolver.php

```diff
@@ -13,37 +13,59 @@
 
 namespace Sylius\Component\Core\Resolver;
 
-use Sylius\Component\Core\Model\ChannelInterface;
 use Sylius\Component\Core\Model\PaymentInterface;
 use Sylius\Component\Core\Repository\PaymentMethodRepositoryInterface;
 use Sylius\Component\Payment\Exception\UnresolvedDefaultPaymentMethodException;
 use Sylius\Component\Payment\Model\PaymentInterface as BasePaymentInterface;
 use Sylius\Component\Payment\Model\PaymentMethodInterface;
 use Sylius\Component\Payment\Resolver\DefaultPaymentMethodResolverInterface;
+use Sylius\Component\Payment\Resolver\PaymentMethodsResolverInterface;
 use Webmozart\Assert\Assert;
 
 class DefaultPaymentMethodResolver implements DefaultPaymentMethodResolverInterface
 {
     /** @var PaymentMethodRepositoryInterface */
     protected $paymentMethodRepository;
 
-    public function __construct(PaymentMethodRepositoryInterface $paymentMethodRepository)
-    {
+    /** @var PaymentMethodsResolverInterface|null */
+    private $paymentMethodsResolver;
+
+    public function __construct(
+        PaymentMethodRepositoryInterface $paymentMethodRepository,
+        ?PaymentMethodsResolverInterface $paymentMethodsResolver = null
+    ) {
         $this->paymentMethodRepository = $paymentMethodRepository;
+
+        if (null === $paymentMethodsResolver) {
+            @trigger_error(
+                sprintf(
+                    'Not passing an $paymentMethodsResolver to "%s" constructor is deprecated since Sylius 1.8 and will be impossible in Sylius 2.0.',
+                    self::class
+                ),
+                \E_USER_DEPRECATED
+            );
+        }
+
+        $this->paymentMethodsResolver = $paymentMethodsResolver;
     }
 
     /**
+     * @param BasePaymentInterface|PaymentInterface $payment
+     *
      * @throws UnresolvedDefaultPaymentMethodException
      */
-    public function getDefaultPaymentMethod(BasePaymentInterface $subject): PaymentMethodInterface
+    public function getDefaultPaymentMethod(BasePaymentInterface $payment): PaymentMethodInterface
     {
-        /** @var PaymentInterface $subject */
-        Assert::isInstanceOf($subject, PaymentInterface::class);
+        Assert::isInstanceOf($payment, PaymentInterface::class);
+
+        $channel = $payment->getOrder()->getChannel();
 
-        /** @var ChannelInterface $channel */
-        $channel = $subject->getOrder()->getChannel();
+        if (null !== $this->paymentMethodsResolver) {
+            $paymentMethods = $this->paymentMethodsResolver->getSupportedMethods($payment);
+        } else {
+            $paymentMethods = $this->paymentMethodRepository->findEnabledForChannel($channel);
+        }
 
-        $paymentMethods = $this->paymentMethodRepository->findEnabledForChannel($channel);
         if (empty($paymentMethods)) {
```

Comment:
```suggestion
        if (!$paymentMethods) {
```

Reply:
Maybe `count($paymentMethods) === 0`?
Or use `reset($paymentMethods)` and check if it's `false` instead of relying on the numeric index (`$paymentMethods[0]`).